### ISSUE TITLE
Version 2.0.0.prerelease.2 dependency issues

### ISSUE THREAD

**Issue by Bobbias** [BindBC/bindbc-sdl#69]
Version 2.0.0.prerelease.2 declares bindbc-common ~>1 as a dependency, but actually depends on changes only present in bindbc-common 1.0.1. Since 1.0.1 is not currently published in the dub registry, it's not currently possible to use 2.0.0.prerelease.2 without explicitly cloning the repo and using `dub add-local <path to bindbc-common>`.

This results in these errors when building (using dynamic bindings and `SDL_MainUseCallbacks`:
```
    Starting Performing "debug" build using D:\DLang\dmd2\windows\bin64\dmd.exe for x86_64.
    Building bindbc-common 1.0.0: building configuration [noBC]
    Building bindbc-loader 1.1.5: building configuration [noBC]
    Building bindbc-sdl 2.0.0-prerelease.2: building configuration [dynamic]
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\bindbc\sdl\config.d(10,15): Error: module `bindbc.common.types` import `c_longlong` not found
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\bindbc\sdl\config.d(10,15): Error: module `bindbc.common.types` import `c_ulonglong` not found
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(63,3): Error: basic type expected, not `=`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(63,5): Error: found `SDL_HapticFeature` when expecting `,`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(63,31): Error: no identifier for declarator `.constant`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(64,3): Error: basic type expected, not `=`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(64,5): Error: found `SDL_HapticFeature` when expecting `,`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(64,27): Error: no identifier for declarator `.sine`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(65,3): Error: basic type expected, not `=`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(65,5): Error: found `SDL_HapticFeature` when expecting `,`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(65,29): Error: no identifier for declarator `.square`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(66,3): Error: basic type expected, not `=`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(66,5): Error: found `SDL_HapticFeature` when expecting `,`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(66,31): Error: no identifier for declarator `.triangle`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(67,3): Error: basic type expected, not `=`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(67,5): Error: found `SDL_HapticFeature` when expecting `,`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(67,33): Error: no identifier for declarator `.sawtoothUp`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(68,3): Error: basic type expected, not `=`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(68,5): Error: found `SDL_HapticFeature` when expecting `,`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\haptic.d-mixin-39(68,35): Error: no identifier for declarator `.sawtoothDown`
```

After cloning bindbc-common and adding it, I get these errors:
```
    Starting Performing "debug" build using D:\DLang\dmd2\windows\bin64\dmd.exe for x86_64.
    Building bindbc-common 1.0.1: building configuration [noBC]
    Building bindbc-loader 1.1.5: building configuration [noBC]
    Building bindbc-sdl 2.0.0-prerelease.2: building configuration [dynamic]
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(54,68): Error: function `_beginthreadex` is not callable using argument types `()`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(54,68):        too few arguments, expected 6, got 0
2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(41,11):        `sdl.thread._beginthreadex(void*, uint, extern (Windows) uint function(void*) nothrow @nogc, void*, uint, uint*)` declared here
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(54,118): Error: function `_endthreadex` is not callable using argument types `()`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(54,118):        too few arguments, expected 1, got 0
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(42,9):        `sdl.thread._endthreadex(uint)` declared here
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(56,73): Error: function `_beginthreadex` is not callable using argument types `()`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(56,73):        too few arguments, expected 6, got 0
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(41,11):        `sdl.thread._beginthreadex(void*, uint, extern (Windows) uint function(void*) nothrow @nogc, void*, uint, uint*)` declared here
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(56,123): Error: function `_endthreadex` is not callable using argument types `()`
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(56,123):        too few arguments, expected 1, got 0
bindbc-sdl\2.0.0-prerelease.2\bindbc-sdl\source\sdl\thread.d(42,9):        `sdl.thread._endthreadex(uint)` declared here
```

Further information: Running windows 11, compiling with DMD version 2.109.1 on Windows 11.

I'm trying to use my QOI image decoder to draw to a single surface, upload that as a texture and just draw the result. Most of my code is taken from the [textures.d](https://github.com/BindBC/bindbc-sdl/blob/master/examples/renderer/textures.d) example.

**Comment by ichordev** [comment]
> Version 2.0.0.prerelease.2 declares bindbc-common ~>1 as a dependency, but actually depends on changes only present in bindbc-common 1.0.1. Since 1.0.1 is not currently published in the dub registry, it's not currently possible to use 2.0.0.prerelease.2 without explicitly cloning the repo and using `dub add-local <path to bindbc-common>`.

I've manually updated the package's releases now. Note that the dub package repository is *supposed* to update the releases automatically, but it takes so long that I always do it manually, and you can't guarantee what order it'll do those updates in. This time I must've forgotten to, so I'm sorry about that.

> After cloning bindbc-common and adding it, I get these errors:
> [...]
> Further information: Running windows 11, compiling with DMD version 2.109.1 on Windows 11.

Ah, that'll be because that's specific to Microsoft platforms. I had to put some random static bindings to `_beginthreadex` and `_endthreadex` (system functions that aren't available in `core.sys.windows`), but then I accidentally aliased them instead of creating pointers to them. 🤦‍♀️

I've pushed & tagged a new release. Try upgrading and let me know if there's any other problems! :)

**Comment by Bobbias** [comment]
Oh wow, thanks for the fast response.

After dealing with the fact that my library is not nothrow, I'm getting a ton of link errors.
```
lld-link: warning: found wmain and wWinMain; defaulting to /subsystem:console
```
Adding `/subsystem:windows` will select the `wWinMain` entrypoint, but it doesn't fix any of the other issues.
```
lld-link: error: <root>: undefined symbol: wmainCRTStartup
lld-link: error: undefined symbol: _fltused
lld-link: error: undefined symbol: printf
lld-link: error: undefined symbol: _tls_index
lld-link: error: undefined symbol: CreateFileW
lld-link: error: undefined symbol: CloseHandle
lld-link: error: undefined symbol: GetFileSize
lld-link: error: undefined symbol: ReadFile
lld-link: error: undefined symbol: FlushFileBuffers
lld-link: error: undefined symbol: GetLastError
lld-link: error: undefined symbol: GetModuleHandleA
lld-link: error: undefined symbol: GetProcAddress
lld-link: error: undefined symbol: LockFileEx
lld-link: error: undefined symbol: UnlockFileEx
lld-link: error: undefined symbol: LoadLibraryA
lld-link: error: undefined symbol: GetCurrentProcess
lld-link: error: undefined symbol: InitializeCriticalSection
lld-link: error: undefined symbol: QueryPerformanceFrequency
lld-link: error: undefined symbol: GetCurrentThreadId
lld-link: error: undefined symbol: GetCurrentThread
lld-link: error: undefined symbol: DuplicateHandle
lld-link: error: undefined symbol: EnterCriticalSection
lld-link: error: undefined symbol: LeaveCriticalSection
lld-link: error: undefined symbol: GetConsoleScreenBufferInfo
lld-link: error: undefined symbol: fprintf
lld-link: error: undefined symbol: LoadLibraryW
lld-link: error: undefined symbol: FreeLibrary
lld-link: error: undefined symbol: GetConsoleOutputCP
lld-link: error: undefined symbol: WideCharToMultiByte
lld-link: error: undefined symbol: WriteFile
lld-link: error: undefined symbol: MultiByteToWideChar
lld-link: error: undefined symbol: DeleteCriticalSection
lld-link: error: undefined symbol: snprintf
lld-link: error: undefined symbol: CreateFileMappingW
lld-link: error: undefined symbol: MapViewOfFileEx
lld-link: error: undefined symbol: FlushViewOfFile
lld-link: error: undefined symbol: UnmapViewOfFile
lld-link: error: undefined symbol: ResumeThread
lld-link: error: undefined symbol: WaitForSingleObject
lld-link: error: undefined symbol: GetThreadPriority
lld-link: error: undefined symbol: SetThreadPriority
lld-link: error: undefined symbol: GetExitCodeThread
lld-link: error: undefined symbol: Sleep
lld-link: error: undefined symbol: SwitchToThread
lld-link: error: undefined symbol: SetEnvironmentVariableW
lld-link: error: undefined symbol: GetEnvironmentStringsW
lld-link: error: undefined symbol: FreeEnvironmentStringsW
lld-link: error: undefined symbol: SetLastError
lld-link: error: undefined symbol: GetEnvironmentVariableW
lld-link: error: undefined symbol: GetExitCodeProcess
lld-link: error: undefined symbol: GetEnvironmentVariableA
lld-link: error: undefined symbol: RtlCaptureContext
lld-link: error: undefined symbol: VirtualAlloc
lld-link: error: undefined symbol: VirtualFree

**Comment by ichordev** [comment]
> DMD is now complaining about not being able to find `Files.d`, but I _think_ that's referring to a file from dlangui, which I had been using previously. I'm not sure how to find out _why_ it's looking for that file. Adding `-v` shows no reference to that file in the output. I'm not sure why it's looking for that file though.

Are you sure the error is from dmd, or is it actually from lld-link?
I never capitalise my file names and there’s never been a module named files in BindBC-SDL, so unfortunately I think your problem is not from BindBC-SDL, meaning I have very little insight into what’s going wrong. Look through dlangui’s dub recipe perhaps?

> And I assume adding those link flags directly rather than adding them via dflags is a suboptimal choice.

You should usually add linker flags via `lflags`. If you want to link a library, add it to `libs`.

> It could also be something weird in the command line where it's interpreting a path with a space in it poorly, but I haven't found where that might be happening.

I had that happen before when I was splitting `-framework CoreFoundation` into two `lflags` items.

Good luck diagnosing the issue! If you get stuck, you could try making a simplified version of your project with fewer dependencies and adding/removing dependencies from there to see how & why it breaks. Also, remember to always `dub upgrade`! ;)

**Comment by Bobbias** [comment]
I created a completely new project and copied my source over. I have 2 files, qoi.d and app.d, so it's already pretty minimal, but doing this would ensure I was starting without any potential leftover configuration issues.

It appears that much of the issues stem from the fact that everything I had been compiling up to this point was compiling as a console application and using all the default entrypoint and runtime setup stuff that D does for you by default. I was also using LLD because I was running my dub commands from the vscode terminal which wasn't set up to use a visual studio developer console (though I have visual studio 2022 installed).

I resolved the linker issues by manually adding a bunch of lib references, so my dflags looks like this:
```json
"dflags": [
    "-L/SUBSYSTEM:WINDOWS",
    "-Luser32.lib",
    "-Lucrt.lib",
    "-Lvcruntime.lib",
    "-Lmsvcrt.lib",
    "-Lkernel32.lib",
    "-Loldnames.lib",
    "-mixin=expanded.mixins",
    "-L/LIBPATH:\"D:\\DLang\\dmd2\\windows\\lib64\""
],
```

However now when I try to run the resulting application, `SDL_RunApp` ends up throwing an access violation. More precisely it's trying to call a null pointer. 

I'm attempting to use the modern SDL callback approach, so I've set `SDL_MainUseCallbacks` in my versions. I have a slightly modified version of the [textures.d](https://github.com/BindBC/bindbc-sdl/blob/master/examples/renderer/textures.d) example where instead of drawing multiple moving textures loaded from a bitmap I'm loading a single QOI file and rendering that to a single texture, with the window set to the size of the texture. All told my `app.d` is 151 lines total.

I don't fully understand exactly what calls are made between the app entrypoint and when `SDL_AppInit` is supposed to be called, so I'd appreciate some insight into what might being going wrong here.

**Comment by ichordev** [comment]
> I created a completely new project and copied my source over. I have 2 files, qoi.d and app.d, so it's already pretty minimal, but doing this would ensure I was starting without any potential leftover configuration issues.
> 
> It appears that much of the issues stem from the fact that everything I had been compiling up to this point was compiling as a console application and using all the default entrypoint and runtime setup stuff that D does for you. I was also using LLD because I was running my dub commands from the vscode terminal which wasn't set up to use a visual studio developer console (though I have visual studio 2022 installed).
> 
> I resolved the linker issues by manually adding a bunch of lib references and running the commands in a visual studio developer console, so it uses `link.exe`, so my current dflags looks like this:
> 
> ```json
> "dflags": [
>     "-L/SUBSYSTEM:WINDOWS",
>     "-Luser32.lib",
>     "-Lucrt.lib",
>     "-Lvcruntime.lib",
>     "-Lmsvcrt.lib",
>     "-Lkernel32.lib",
>     "-Loldnames.lib",
>     "-mixin=expanded.mixins",
>     "-L/LIBPATH:\"D:\\DLang\\dmd2\\windows\\lib64\""
> ],
> ```

Yeah, like the readme explains: using  callback functions mean your main has to be `extern(C) nothrow` so that SDL can call it; which bypasses all of DRuntime’s usual done-for-you setup. I guess that includes linking to Windows system libraries well for some reason? I didn’t have any problems like that on macOS.

> I'm attempting to use the modern SDL callback approach, so I've set `SDL_MainUseCallbacks` in my versions. I have a slightly modified version of the [textures.d](https://github.com/BindBC/bindbc-sdl/blob/master/examples/renderer/textures.d) example where instead of drawing multiple moving textures loaded from a bitmap I'm loading a single QOI file and rendering that to a single texture, with the window set to the size of the texture. All told my `app.d` is 151 lines total.

I’m guessing you don’t happen to have a public repository for it?

> I don't fully understand exactly what calls are made between the app entrypoint and when `SDL_AppInit` is supposed to be called

https://github.com/BindBC/bindbc-sdl/blob/41b1d0f03d54ee4dc8875df31388a836d91a4263/source/sdl/main.d#L58

TL;DR it just calls `SDL_RunApp`, which calls `SDL_main` somehow, which calls `EnterAppMainCallbacks`, which contains platform-specific code that runs your callbacks.

> Looking at things in WinDBG, This line is where the exception occurs:
> 
> ```
> 00007ff7`905d6426 48ff154b4a1800   call    qword ptr [d_qoi2!sdl.main._SDL_RunApp (7ff79075ae78)]
> ```
> 
> The address it shows here for `_SDL_RunApp`: `7ff79075ae78` is null, but if I look at the "locals" section, mainFunction contains the address `0x00007ff790586433`, which is part of the ILT:
> 
> ```
>     d_qoi2!ILT+21550(SDL_main):
> 00007ff7`90586433 e958bb0000 jmp     d_qoi2!app.SDL_main (7ff790591f90)
> ```
> 
> Following that address we get the correct `SDL_main` function:
> 
> ```
>     d_qoi2!app.SDL_main:
> 00007ff7`90591f90 55               push    rbp
> 00007ff7`90591f91 488bec           mov     rbp, rsp
> 00007ff7`90591f94 894d10           mov     dword ptr [argC (rbp+10h)], ecx
> 00007ff7`90591f97 48895518         mov     qword ptr [argV (rbp+18h)], rdx
> 00007ff7`90591f9b 488d056624ffff   lea     rax, [d_qoi2!@ILT+13315(SDL_AppQuit) (7ff790584408)]
> 00007ff7`90591fa2 50               push    rax
> 00007ff7`90591fa3 488d0d3ef9feff   lea     rcx, [d_qoi2!@ILT+2275(SDL_AppEvent) (7ff7905818e8)]
> 00007ff7`90591faa 51               push    rcx
> 00007ff7`90591fab 4c8d0dbbfbfeff   lea     r9, [d_qoi2!@ILT+2920(SDL_AppIterate) (7ff790581b6d)]
> 00007ff7`90591fb2 4c8d05c91dffff   lea     r8, [d_qoi2!@ILT+11645(SDL_AppInit) (7ff790583d82)]
> 00007ff7`90591fb9 488b5518         mov     rdx, qword ptr [argV (rbp+18h)]
> 00007ff7`90591fbd 8b4d10           mov     ecx, dword ptr [argC (rbp+10h)]
> 00007ff7`90591fc0 4883ec20         sub     rsp, 20h
> 00007ff7`90591fc4 e8c156ffff       call    d_qoi2!@ILT+26245(SDL_EnterAppMainCallbacks) (7ff79058768a)
> 00007ff7`90591fc9 4883c430         add     rsp, 30h
> 00007ff7`90591fcd 5d               pop     rbp
> 00007ff7`90591fce c3               ret     
> ```
> 
> So I'm rather confused. For reference, this is `_SDL_RunApp`:
> 
> ```
>    d_qoi2!sdl.main._SDL_RunApp:
> 00007ff7`9075ae78 0000 add     byte ptr [rax], al
> 00007ff7`9075ae7a 0000 add     byte ptr [rax], al
> 00007ff7`9075ae7c 0000 add     byte ptr [rax], al
> 00007ff7`9075ae7e 0000 add     byte ptr [rax], al
> ```
> 
> So why is `_SDL_RunApp` null?

Are you using static or dynamic bindings? If you’re using dynamic, what does your dynLoad code look like? If loading SDL3 in the dynamic bindings fails, SDL_RunApp might not be loaded. Otherwise, the OS would be loading SDL_RunApp and it should give you an error message if it fails.

**Comment by Bobbias** [comment]
> Yeah, like the readme explains: using callback functions mean your main has to be `extern(C) nothrow` so that SDL can call it; which bypasses all of DRuntime’s usual done-for-you setup. I guess that includes linking to Windows system libraries well for some reason? I didn’t have any problems like that on macOS.

From the reading I've done I get the feeling that writing programs that use the `WinMain` entrypoint (or `wWinMain`) requires a bunch of manual setup, but I haven't tried that in isolation.

> I’m guessing you don’t happen to have a public repository for it?

Not yet, though I I do plan to put it up after I've cleaned things up and made sure I'm actually getting pixel perfect decoding.

> TL;DR it just calls `SDL_RunApp`, which calls `SDL_main` somehow, which calls `EnterAppMainCallbacks`, which contains platform-specific code that runs your callbacks.

Ok, so basically only one step in the chain seems broken.

> Are you using static or dynamic bindings? If you’re using dynamic, what does your dynLoad code look like? If loading SDL3 in the dynamic bindings fails, SDL_RunApp might not be loaded. Otherwise, the OS would be loading SDL_RunApp and it should give you an error message if it fails.

I'm using dynamic bindings:
```D
extern (C) nothrow:
mixin(makeSDLMain(dynLoad: q{
    if(!loadSDL()){
        import core.stdc.stdio, bindbc.loader;
        foreach(error; bindbc.loader.errors){
            printf("%s\n", error.message);
        }
    }}));
```

I've thrown what I've got so far into a [gist here](https://gist.github.com/Bobbias/3f0de5c0f72c2f342fabe27a610b3cfa), though I didn't bother adding a dub.json, since it appears that basically everything I've had to do is windows specific. There are some Versions and dependencies specific to my libray (some unit tests depend on `arsd-official:png`, but the library itself does not).

**Comment by ichordev** [comment]
Sorry for the late response.

> From the reading I've done I get the feeling that writing programs that use the `WinMain` entrypoint (or `wWinMain`) requires a bunch of manual setup, but I haven't tried that in isolation.

I thought SDL was meant to handle most of that in SDL_RunApp. Maybe I’m wrong.

> I'm using dynamic bindings

They’re certainly more error-prone, so that’s a good clue.

> ```d
> extern (C) nothrow:
> mixin(makeSDLMain(dynLoad: q{
>     if(!loadSDL()){
>         import core.stdc.stdio, bindbc.loader;
>         foreach(error; bindbc.loader.errors){
>             printf("%s\n", error.message);
>         }
>     }}));
> ```

Ah, I made an error here. You see, `loadSDL` now returns an enum `bindbc.loader.sharedlib.LoadMsg` where ‘zero’ is success. So ‘if not’ is basically ‘if loaded correctly’, whoops! You actually want `if(loadSDL() != LoadMsg.success)`, or to avoid an import: `loadSDL(); if(isSDLLoaded())` (where the latter is probably more future-proof)
Sorry about that. This probably means that you’re having some sort of runtime linking error. (e.g. your system doesn’t know where SDL3 is)

**Comment by Bobbias** [comment]
No problem, this isn't any sort of high priority project. And yeah, I actually ended up discovering the `LoadMsg` inverted logic part out on my own. Once I realized that I didn't have SDL in the folder with the application and tossed it in there, because I had added an explicit return error code inside that if statement I could see it was now entering the if body and hitting my return statement. From there I looked into the source for bindbc loader and figured out that the logic in that check was reversed.

Everything loads fine now.

I would suggest updating [readme.md#replacing-the-main-entry-point](https://github.com/BindBC/bindbc-sdl?tab=readme-ov-file#replacing-the-main-entry-point) and the `Callback entry points` code to reflect this, because that's where I copied my initial code from. I noticed that the code directly above the `Binding-specific changes` section is updated so it's only those parts that are out of date.

**Comment by ichordev** [comment]
> No problem, this isn't any sort of high priority project. And yeah, I actually ended up discovering the `LoadMsg` inverted logic part out on my own. Once I realized that I didn't have SDL in the folder with the application and tossed it in there, because I had added an explicit return error code inside that if statement I could see it was now entering the if body and hitting my return statement. From there I looked into the source for bindbc loader and figured out that the logic in that check was reversed.
> 
> Everything loads fine now.
> 
> I would suggest updating [readme.md#replacing-the-main-entry-point](https://github.com/BindBC/bindbc-sdl?tab=readme-ov-file#replacing-the-main-entry-point) and the `Callback entry points` code to reflect this, because that's where I copied my initial code from. I noticed that the code directly above the `Binding-specific changes` section is updated so it's only those parts that are out of date.
> 
> In any case I suppose this issue can be closed at this point. All my remaining issues are purely issues in my own code. For example, a windows application does not get stdin/out/err, so writeln() logging breaks, and I was being lazy and using that all over my library.

Fantastic that we were able to resolve this, and thank you so much for your cooperation! Usually I stress more about making the readme correct before releasing a BindBC project, but this is a pre-release so I figured I'd just get people using it sooner to find bugs. In the end you uncovered a mistake that I wouldn't have noticed even after double-checking, so thank you again! :) I will update that readme to reflect the changed API.
